lib/Sema/CSStep.h

```diff
@@ -294,6 +294,11 @@ class DependentComponentSplitterStep final : public SolverStep {
   /// Array containing all of the partial solutions for the parent split.
   MutableArrayRef<SmallVector<Solution, 4>> AllPartialSolutions;
 
+  /// The solutions computed the \c ComponentSteps created for each partial
+  /// solution combinations. Will be merged into the final \c Solutions vector
+  /// in \c resume.
+  SmallVector<std::unique_ptr<SmallVector<Solution, 2>>, 2> ContextualSolutions;
```

Comment:
Should we consider using plain std::vector here instead of wasting stack?

Reply:
I don’t have strong opinions but I think based on the fact that all but two tests passed with the `SmallVector` of size 2, it’s reasonable to assume that there are often only `ComponentSteps` that need to be taken here, so I think it’s worth to avoid the heap allocation of `std::vector` here, especially since the member of `ContextualSolutions` is only a pointer and thus not really occupying a whole lot of stack space.

Reply:
We just might not have too many examples of complex result builders in the suite., but I am okay with leaving it as small vector for now too.

Reply:
Hmm, that’s true. I still think there’s still value in using the `SmallVector` and the overhead should be relatively low.

Reply:
The only concern I have is for sufficiently complex result builders not to run out of stack in memory constrained environments  because each dependent component would now allocate this storage and keep it along the whole duration of sub-path processing in splitter step.

Reply:
Maybe I’m missing something but I thought the constraint solver had a very low stack usage because it is using a work list to schedule new steps. 

But if you think that’s an issue, I’ll change it to `std::vector`. As I said, I don’t have strong feelings about this.

Reply:
Constraint system and solutions use plenty of stack, we had to convert some of the small vectors to std::vector to mitigate that.